### PR TITLE
[FSTORE-1285] Model Dependent Transformation Functions

### PR DESCRIPTION
**Pull request by manu-sj**
This PR implements model dependent transformation function as per the design log : https://hopsworks.atlassian.net/wiki/spaces/FST/pages/388464653/Model+Dependent+Transformation+Proposal+-+version+2

**Changes Performed**
- Added decorator called `hopsworks_udf` that should be used to defined transformation functions.
- Added possibility to obtain statistics within transformation functions.
- Added possibility for one to many and many to many transformation functions.
- Transformation functions executed as `pandas_udf's` in spark engine and regular pandas functions in python engine.

- Added ability to attach to a feature view transformation functions that are not explicitly saved to the backend using the transformations API.
- Removed attaching transformation function to training dataset features and attached transformation functions to feature view directly. Hence removed the DTO class `transformation_functions_attached`.

JIRA Issue: https://hopsworks.atlassian.net/browse/FSTORE-1285?atlOrigin=eyJpIjoiYjYwZDAyM2ZjZjNkNDU5MWEyNmRjN2MxZjNiZmNhZmEiLCJwIjoiaiJ9

Related PRs: 
- https://github.com/logicalclocks/hopsworks-ee/pull/1775
- https://github.com/logicalclocks/hopsworks-chef/pull/1140
- https://github.com/logicalclocks/loadtest/pull/370

**How Has This Been Tested?**

- [x] Unit Tests
- [x] Integration Tests
- [x] Manual Tests on VM

**Checklist For The Assigned Reviewer:**

```
- [ ] Checked if merge conflicts with master exist
- [ ] Checked if stylechecks for Java and Python pass
- [ ] Checked if all docstrings were added and/or updated appropriately
- [ ] Ran spellcheck on docstring
- [ ] Checked if guides & concepts need to be updated
- [ ] Checked if naming conventions for parameters and variables were followed
- [ ] Checked if private methods are properly declared and used
- [ ] Checked if hard-to-understand areas of code are commented
- [ ] Checked if tests are effective
- [ ] Built and deployed changes on dev VM and tested manually
- [x] (Checked if all type annotations were added and/or updated appropriately)
```

### CODE REVIEW

File: python/hsfs/feature_view.py [hunk 1]
```diff
@@ -56,9 +54,11 @@
 from hsfs.core.vector_db_client import VectorDbClient
 from hsfs.decorators import typechecked
 from hsfs.feature import Feature
+from hsfs.hopsworks_udf import HopsworksUdf
 from hsfs.statistics import Statistics
 from hsfs.statistics_config import StatisticsConfig
 from hsfs.training_dataset_split import TrainingDatasetSplit
+from hsfs.transformation_function import TransformationFunction
```
Comment (kennethmhc) [inline python/hsfs/feature_view.py:61]: will be nice to add a property which return the list of feature name after transformation

File: python/hsfs/engine/python.py [hunk 2]
```diff
@@ -1228,39 +1292,59 @@ def add_file(self, file: Optional[str]) -> Optional[str]:
...
+        # Raises
+            `FeatureStoreException`: If any of the features mentioned in the transformation function is not present in the Feature View.
+        """
+        transformed_features = set()
+
+        if isinstance(dataset, pl.DataFrame) or isinstance(
+            dataset, pl.dataframe.frame.DataFrame
+        ):
+            # Converting polars dataframe to pandas because currently we support only pandas UDF's as transformation functions.
+            if os.getenv("USE_PYARROW_EXTENSION", False):
```
Comment (kennethmhc) [inline python/hsfs/engine/python.py:1315]: seems not necessary to use env var here, just a method is sufficient as you do not expect user to set it.

File: python/hsfs/engine/spark.py [hunk 3]
```diff
@@ -1162,67 +1228,68 @@ def add_cols_to_delta_table(self, feature_group, new_features):
...
+
+            pandas_udf = hopsworks_udf.get_udf()
+            output_col_name = hopsworks_udf.output_column_names[0]
+
+            transformations.append(pandas_udf)
+            output_col_names.append(output_col_name)
+            transformation_features.append(hopsworks_udf.transformation_features)
+
+            if len(hopsworks_udf.output_types) > 1:
+                explode_name.append(f"{output_col_name}.*")
```
[the remark this answers is not in the document]
Comment (manu-sj) [reply python/hsfs/engine/spark.py:1273]: All columns names are maintained as a list to preserve the order of elements. 

The Dataframe after transformation would have the untransformed elements in the same order as in the input dataframe and then have the transformed features. 

The order of transformed features in the output dataframe would is made consistent by sorted in transformation function based on the output_column_name when they are attached to the feature view. 

So the transformed feature in the output dataframe has ordered by sorting it based on the output column name/

Also the unpacking operator * maintains the ordering of the list - https://stackoverflow.com/questions/51435635/is-args-guaranteed-to-preserve-order

File: python/hsfs/feature_view.py (hunk 1 above)
Comment (manu-sj) [reply python/hsfs/feature_view.py:61]: added property `output_column_names`

File: python/hsfs/engine/python.py (hunk 2 above)
Comment (manu-sj) [reply python/hsfs/engine/python.py:1315]: Changed to a function.

File: python/hsfs/feature_view.py (hunk 1 above)
Comment (kennethmhc) [reply python/hsfs/feature_view.py:61]: I mean `fv.transformed_features`, maybe a better name, should return the list of feature name you will get in `get_feature_vector`

Comment (manu-sj) [reply python/hsfs/feature_view.py:61]: Renamed property that returns list of features names in a feature view after transformation to `transformed_features` .

Comment (manu-sj) [reply python/hsfs/feature_view.py:61]: Added property transformed_features in feature view to get name of features after transformation functions have been applied.

**Review by kennethmhc — APPROVED** [review]
LGTM!

File: python/hsfs/transformation_statistics.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        json_decamelized = humps.decamelize(json_dict)
+        return cls(**json_decamelized)
+
+
+class TransformationStatistics:
+    """
+    Class that stores statistics of all features required for a transformation function.
+    """
+
+    def __init__(self, *features):
```
Comment (kennethmhc) [inline python/hsfs/transformation_statistics.py:109]: missing type hint

File: python/hsfs/transformation_statistics.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        self._features = features
+        self.__dict__.update(
+            {feature: self.init_statistics(feature) for feature in features}
+        )
+
+    def init_statistics(self, feature_name):
```
Comment (kennethmhc) [inline python/hsfs/transformation_statistics.py:115]: missing type hint.

**Review by kennethmhc — CHANGES_REQUESTED** [review]

File: python/hsfs/transformation_statistics.py (hunk 5 above)
Comment (manu-sj) [reply python/hsfs/transformation_statistics.py]: added type hints

File: python/hsfs/transformation_statistics.py (hunk 4 above)
Comment (manu-sj) [reply python/hsfs/transformation_statistics.py]: added type hints

**Review by kennethmhc — APPROVED** [review]

**Pull request merged**